credit/nwp.py

```diff
@@ -0,0 +1,113 @@
+import numpy as np
+import pandas as pd
+from os.path import join
+import xarray as xr
+import fsspec
+import xesmf as xe
+
+gfs_map = {'tmp': 'T', 'ugrd': 'U', 'vgrd': 'V', 'spfh': 'Q', 'pressfc': 'SP', 'tmp2m': 't2m', 'delz': 'Z500'}
+level_map = {'T500': 'T', 'U500': 'U', 'V500': 'V', 'Q500': 'Q', 'Z500': 'Z'}
+upper_air = ['T', 'U', 'V', 'Q', 'Z']
+surface = ['SP', 't2m']
+
+def build_GFS_init(output_grid, gdas_base_path="https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/",
+                   date="202502200600", variables=[], model_level_indices=[]):
+
+    gfs_variables = [k for k, v in gfs_map.items() if v in variables]
+    atm_full_path = build_file_path(date, gdas_base_path, file_type='atm')
+    sfc_full_path = build_file_path(date, gdas_base_path, file_type='sfc')
+    gfs_atm_data = load_gfs_data(atm_full_path, gfs_variables)
+    gfs_sfc_data = load_gfs_data(sfc_full_path, gfs_variables)
+    gfs_data = combine_data(gfs_atm_data, gfs_sfc_data)
+    regridded_gfs = regrid(gfs_data, output_grid)
+    interpolated_gfs = interpolate_to_model_level(regridded_gfs, output_grid, model_level_indices, variables)
+    final_data = format_data(interpolated_gfs, regridded_gfs, model_level_indices)
+
+    return final_data
+
+
+def build_file_path(date, base_path, file_type='atm'):
+
+    date_obj = pd.Timestamp(date)
+    dir_path = date_obj.strftime("gdas.%Y%m%d/%H/atmos/")
+    file_name = date_obj.strftime(f"gdas.t%Hz.{file_type}anl.nc")
+
+    return join(base_path, dir_path, file_name)
+
+
+def load_gfs_data(full_file_path, variables):
+
+    ds = xr.open_dataset(fsspec.open(full_file_path).open())
+    available_vars = ds.data_vars
+    vars = [v for v in variables if v in available_vars]
+    ds = ds[vars].rename({'grid_xt': 'longitude', 'grid_yt': 'latitude'}).load()
+    if 'delz' in available_vars and 'delz' in variables:
+        ds['delz'].values = np.abs(ds['delz'].cumsum(dim='pfull').values) * 9.81
```

Comment:
The summation needs to be done in reverse since the model level closest to the surface has the last index.